File: 206.reverse-linked-list.py
#
# @lc app=leetcode id=206 lang=python3
#
# [206] Reverse Linked List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if head is None:
            return None
        new_head = None
        while head:
            node = head
            head = head.next
            node.next = new_head
            new_head = node
        return new_head
        
# @lc code=end

File: golang/206.reverse-linked-list.go
/*
 * @lc app=leetcode.cn id=206 lang=golang
 *
 * [206] Reverse Linked List
 */
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func reverseList(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}
	newHead, tail := reverseListHelper(head)
	tail.Next = nil
	return newHead
}

func reverseListHelper(head *ListNode) (newHead, tail *ListNode) {
	if head.Next == nil {
		return head, head
	}
	newHead, tail = reverseListHelper(head.Next)
	tail.Next = head
	tail = tail.Next

	return newHead, tail

}

// @lc code=end
